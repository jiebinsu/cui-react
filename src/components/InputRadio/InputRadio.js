import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { useRadioContext } from "../InputRadioGroup/InputRadioGroupContext";
import classNames from "../../utils/classNames";
import "./inputRadio.scss";

const InputRadio = ({
  id,
  name,
  value,
  className,
  label,
  checked,
  value: radioValue,
  disabled,
  onChange,
  ...props
}) => {
  const [selfChecked, setSelfChecked] = useState(!!checked);
  const context = useRadioContext();
  const { currentState, updateState, inGroup } = context;
  const classes = classNames(
    "coop-form__field coop-form__radio coop-c-form-choice__input coop-c-form-choice__input--radio-button",
    [className]
  );

  if (inGroup) {
    useEffect(() => {
      setSelfChecked(currentState === radioValue);
    }, [currentState, radioValue]);
  }

  const onChangeHandler = (event) => {
    if (disabled) return;
    const selfEvent = {
      target: {
        checked: !selfChecked,
      },
      stopPropagation: event.stopPropagation,
      preventDefault: event.preventDefault,
      nativeEvent: event,
    };
    setSelfChecked(!selfChecked);
    if (inGroup) {
      if (updateState) updateState(radioValue);
    }
    if (onChange) onChange(selfEvent);
  };

  useEffect(() => {
    if (checked === undefined) return;
    setSelfChecked(Boolean(checked));
  }, [checked]);

  const tagAttributes = {
    id,
    name,
    value: radioValue,
    label,
    disabled,
    type: "radio",
    className: classes,
    checked: selfChecked,
    onChange: onChangeHandler,
    ...props,
  };

  return (
    <div className="coop-c-form-choice">
      <input {...tagAttributes} />
      <label
        className="coop-form__label coop-c-form-choice__label"
        htmlFor={id}
      >
        {label}
      </label>
    </div>
  );
};

InputRadio.defaultProps = {
  disabled: false,
  className: null,
  checked: null,
  onChange: null,
};

InputRadio.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
  className: PropTypes.string,
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
};

export default InputRadio;
