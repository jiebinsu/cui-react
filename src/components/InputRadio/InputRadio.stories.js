import React from "react";
import InputRadio from "./InputRadio";

export default {
  title: "Base/InputRadio",
  component: InputRadio,
};

export const Basic = (args) => <InputRadio {...args} />;
Basic.args = {
  id: "basic",
  name: "basic",
  label: "Basic Radio",
  value: 1,
};
