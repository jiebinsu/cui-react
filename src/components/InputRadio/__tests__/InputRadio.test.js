import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import InputRadio from "../index";

describe("InputRadio", () => {
  it("should render correctly", () => {
    const wrapper = mount(
      <InputRadio id="test" name="test" label="test" value={1} />
    );
    expect(() => wrapper.unmount()).not.toThrow();

    const tree = renderer
      .create(<InputRadio id="test" name="test" label="test" value={1} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
