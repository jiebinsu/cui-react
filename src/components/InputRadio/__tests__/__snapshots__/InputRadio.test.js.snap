// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputRadio should render correctly 1`] = `
<div
  className="coop-c-form-choice"
>
  <input
    checked={false}
    className="coop-form__field coop-form__radio coop-c-form-choice__input coop-c-form-choice__input--radio-button"
    disabled={false}
    id="test"
    name="test"
    onChange={[Function]}
    type="radio"
    value={1}
  />
  <label
    className="coop-form__label coop-c-form-choice__label"
    htmlFor="test"
  >
    test
  </label>
</div>
`;
